language/tools/test_generation/src/summaries.rs

```diff
@@ -0,0 +1,262 @@
+    state_stack_push_local_copy, state_stack_push_local_move, transitions::*,
+};
+use vm::file_format::{Bytecode, SignatureToken};
+
+/// A `Precondition` is a boolean predicate on an `AbstractState`.
+type Precondition = dyn Fn(&AbstractState) -> bool;
+
+/// A `Effect` is a function that transforms on `AbstractState` to another
+type Effect = dyn Fn(&AbstractState) -> AbstractState;
+
+/// The `Summary` of a bytecode instruction contains a list of `Precondition`s
+/// and a list of `Effect`s.
+pub struct Summary {
+    pub preconditions: Vec<Box<Precondition>>,
+    pub effects: Vec<Box<Effect>>,
+}
+
+/// Return the `Summary` for a bytecode instruction, `instruction`
+pub fn instruction_summary(instruction: Bytecode) -> Summary {
+    match instruction {
+        Bytecode::Pop => Summary {
+            preconditions: vec![state_stack_has!(0, None)],
+            effects: vec![state_stack_pop!()],
+        },
+        Bytecode::LdConst(_) => Summary {
+            preconditions: vec![],
+            effects: vec![state_stack_push!(SignatureToken::U64)],
+        },
+        Bytecode::LdStr(_) => Summary {
+            preconditions: vec![],
+            effects: vec![state_stack_push!(SignatureToken::String)],
+        },
+        Bytecode::LdAddr(_) => Summary {
+            preconditions: vec![],
+            effects: vec![state_stack_push!(SignatureToken::Address)],
+        },
+        Bytecode::LdTrue => Summary {
+            preconditions: vec![],
+            effects: vec![state_stack_push!(SignatureToken::Bool)],
+        },
+        Bytecode::LdFalse => Summary {
+            preconditions: vec![],
+            effects: vec![state_stack_push!(SignatureToken::Bool)],
+        },
+        Bytecode::CopyLoc(i) => Summary {
+            preconditions: vec![state_local_available!(i)],
+            effects: vec![state_stack_push_local_copy!(i)],
+        },
+        Bytecode::MoveLoc(i) => Summary {
+            preconditions: vec![state_local_available!(i)],
+            effects: vec![state_stack_push_local_move!(i)],
+        },
+        Bytecode::StLoc(i) => Summary {
+            preconditions: vec![state_stack_has!(0, None), state_local_available!(i)],
+            effects: vec![state_stack_pop_local_insert!(i), state_stack_pop!()],
+        },
+        Bytecode::BorrowLoc(i) => Summary {
+            preconditions: vec![state_local_available!(i)],
+            effects: vec![state_stack_push_local_borrow!(i)],
+        },
+        Bytecode::Add => Summary {
+            preconditions: vec![
+                state_stack_has!(0, Some(SignatureToken::U64)),
+                state_stack_has!(1, Some(SignatureToken::U64)),
+            ],
+            effects: vec![
+                state_stack_pop!(),
+                state_stack_pop!(),
+                state_stack_push!(SignatureToken::U64),
+            ],
+        },
+        Bytecode::Sub => Summary {
```

Comment:
Might be misunderstanding exactly what is happening here, but how do you handle the relationship between `Sub`s arguments? We don't currently allow negative numbers and because of this, the instruction's parameters are somewhat "entangled" and integer arguments cannot be generated independently from each other. Might be worth putting a TODO on this if that's not currently handled.

Reply:
Bytecode generation currently models values abstractly at the level of their type so we wouldn't capture that constraint. I think that enforcing that constraint would require a concrete view of values, which is possible, but not the first priority. I'll add a TODO so that point doesn't get lost.